### ISSUE TITLE
Behaviour when j has undefined columns and i-operation leads to 0-length.

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#1002]
Following #1001 

```R
require(data.table) ## 1.9.5
DT = data.table(a=1:2)
options(datatable.auto.index=FALSE)
DT[a==3, b := d+1L] # no error
#    a
# 1: 1
# 2: 2
setkey(DT, a)
DT[J(3), b := d+1L, nomatch=0L] # error
# Error in eval(expr, envir, enclos) : object 'd' not found
```

I'm not convinced if the behaviour on subset is right.

**Comment by gsee** [comment]
Here is something a little closer to the actual usecase.  I don't think either should be an error, but if the second one is, then the first one should be too, IMO

````
DT[J(3), b := 1L, nomatch=0L] # no error
DT[J(3), d:=b + 2, nomatch=0L] # error
#Error in eval(expr, envir, enclos) : object 'b' not found
````

**Comment by arunsrinivasan** [comment]
No error makes more sense to me too.

**Comment by arunsrinivasan** [comment]
Seems to have been fixed in v1.9.6...

**Comment by arunsrinivasan** [comment]
When `nomatch=NA`, both expressions return an error (as expected). And otherwise, there's no rows to evaluate on. So `j-exp` is not looked at all = no error (as expected).

The error on NA is very likely due to https://github.com/Rdatatable/data.table/issues/759 fix.
